File: sdk/ts/test/conceptBrowser.test.ts
// Concept-browser tests. The query string is the contract: it MUST declare
// sort + paginate, or the engine applies its implicit unmarked-list backstop
// (MEMORY_ENGINE_DEFAULT_LIST_CAP, 50 rows) and silently truncates with no
// continuation cursor (memql#2008). These tests pin the emitted string.

import test from "node:test";
import assert from "node:assert/strict";

import {
  browseConceptPage,
  getRowByConceptAndId,
  DEFAULT_CONCEPT_BROWSE_PAGE_SIZE,
} from "../src/client/conceptBrowser.js";
import { QueryClient } from "../src/client/query.js";
import type { Dispatcher } from "../src/client/dispatcher.js";
import type { ClientMessage, ServerMessage } from "../src/client/wire.js";

class MockDispatcher {
  readonly sent: ClientMessage[] = [];
  private reply: Record<string, unknown> = {};

  setReply(payload: Record<string, unknown>): void {
    this.reply = payload;
  }

  send(msg: ClientMessage): string {
    this.sent.push(msg);
    return "mock-0";
  }

  async sendAndWait(msg: ClientMessage): Promise<ServerMessage> {
    this.sent.push(msg);
    return this.reply as unknown as ServerMessage;
  }

  addEventListener(): () => void {
    return () => {};
  }

  registerStream(): () => void {
    return () => {};
  }

  lastQuery(): string {
    const last = this.sent.at(-1) as { executeQuery?: { query?: string } } | undefined;
    return last?.executeQuery?.query ?? "";
  }

  lastCursor(): string | undefined {
    const last = this.sent.at(-1) as { executeQuery?: { cursor?: string } } | undefined;
    return last?.executeQuery?.cursor;
  }
}

function client(d: MockDispatcher): QueryClient {
  return new QueryClient(d as unknown as Dispatcher);
}

function bundleReply(nodes: unknown[], cursor?: string): Record<string, unknown> {
  return {
    queryResult: {
      requestId: "r",
      result: {
        bundle: { nodes },
        ...(cursor === undefined ? {} : { meta: { cursor } }),
      },
    },
  };
}

test("emits a sort+paginate query so the keyset window applies", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  await browseConceptPage(client(d), "v1:agents:agent");
  assert.equal(
    d.lastQuery(),
    `sort(paginate(concept==v1:agents:agent, ${DEFAULT_CONCEPT_BROWSE_PAGE_SIZE}), "createdAt", "asc")`,
  );
});

test("honors an explicit page size", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  await browseConceptPage(client(d), "v1:agents:agent", { pageSize: 25 });
  assert.match(d.lastQuery(), /paginate\(concept==v1:agents:agent, 25\)/);
});

test("falls back to the default page size for a non-positive value", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  await browseConceptPage(client(d), "v1:agents:agent", { pageSize: 0 });
  assert.match(
    d.lastQuery(),
    new RegExp(`paginate\\(concept==v1:agents:agent, ${DEFAULT_CONCEPT_BROWSE_PAGE_SIZE}\\)`),
  );
});

test("forwards the continuation cursor", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  await browseConceptPage(client(d), "v1:agents:agent", { cursor: "opaque-1" });
  assert.equal(d.lastCursor(), "opaque-1");
});

test("omits the cursor on a first page", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  await browseConceptPage(client(d), "v1:agents:agent");
  assert.equal(d.lastCursor(), undefined);
});

test("preserves the full nested node shape rather than flattening", async () => {
  const d = new MockDispatcher();
  d.setReply(
    bundleReply([
      { id: "a1", concept: "v1:agents:agent", payload: { name: "Sofia" } },
    ]),
  );
  const page = await browseConceptPage(client(d), "v1:agents:agent");
  assert.equal(page.rows.length, 1);
  assert.deepEqual(page.rows[0]?.["payload"], { name: "Sofia" });
});

test("returns the next cursor when the engine mints one", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([{ id: "a1" }], "next-page"));
  const page = await browseConceptPage(client(d), "v1:agents:agent");
  assert.equal(page.nextCursor, "next-page");
});

test("returns an empty cursor when the set is exhausted", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([{ id: "a1" }]));
  const page = await browseConceptPage(client(d), "v1:agents:agent");
  assert.equal(page.nextCursor, "");
});

test("rejects an empty concept id", async () => {
  const d = new MockDispatcher();
  await assert.rejects(
    () => browseConceptPage(client(d), ""),
    /conceptId is required/,
  );
});

test("getRowByConceptAndId queries by concept and id", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([{ id: "a1", payload: { name: "Sofia" } }]));
  const row = await getRowByConceptAndId(client(d), "v1:agents:agent", "a1");
  assert.equal(d.lastQuery(), "concept==v1:agents:agent && id==a1");
  assert.deepEqual(row?.["payload"], { name: "Sofia" });
});

test("getRowByConceptAndId returns null when nothing matches", async () => {
  const d = new MockDispatcher();
  d.setReply(bundleReply([]));
  const row = await getRowByConceptAndId(client(d), "v1:agents:agent", "nope");
  assert.equal(row, null);
});

test("getRowByConceptAndId rejects a missing row id", async () => {
  const d = new MockDispatcher();
  await assert.rejects(
    () => getRowByConceptAndId(client(d), "v1:agents:agent", ""),
    /rowId is required/,
  );
});

File: sdk/go/client/concept_browser_test.go
package client

import (
	"context"
	"testing"
	"time"

	memqlv1 "github.com/znasllc-io/memql/component/grpc/gen"
	"google.golang.org/protobuf/types/known/structpb"
)

// Reuses the mockStream helper from dispatcher_test.go (same package).

// TestBrowseConceptPage_RoundTrip proves the keyset-paginated concept-browse
// wrapper (1) declares sort + paginate so the query opts OUT of the engine's
// implicit 50-row unmarked-list backstop and into a keyset window, (2) stamps
// the inbound continuation cursor on ExecuteQueryMsg.cursor, and (3) returns
// the engine's nextCursor + the page rows (full nested node shape). This is
// the contract the cockpit Concepts tab relies on to page past 50 rows
// (memql#2008).
func TestBrowseConceptPage_RoundTrip(t *testing.T) {
	stream := newMockStream()
	d := NewDispatcher(stream, nil)
	go d.Run()
	defer d.Stop()
	qc := NewQueryClient(d)

	type out struct {
		page *PageResult
		err  error
	}
	done := make(chan out, 1)
	go func() {
		p, err := qc.BrowseConceptPage(context.Background(), "v1:cognition:utterance", "cursor-from-page-1", 200)
		done <- out{page: p, err: err}
	}()

	sent := <-stream.sendCh
	q := sent.GetExecuteQuery()
	if q == nil {
		t.Fatalf("expected ExecuteQueryMsg payload, got %+v", sent.GetPayload())
	}
	wantQuery := `sort(paginate(concept==v1:cognition:utterance, 200), "createdAt", "asc")`
	if q.GetQuery() != wantQuery {
		t.Errorf("query: want %q, got %q (must declare sort+paginate to escape the 50-row backstop)", wantQuery, q.GetQuery())
	}
	if q.GetCursor() != "cursor-from-page-1" {
		t.Errorf("cursor: want %q, got %q (inbound cursor must ride the request)", "cursor-from-page-1", q.GetCursor())
	}

	// Full nested MemoryNode shape (bundle path) -- payload + intrinsics
	// survive, which the cockpit's generic inspector needs.
	payload, _ := structpb.NewStruct(map[string]any{"text": "hello"})
	stream.recvCh <- &memqlv1.MemqlServerMessage{
		CorrelateTo: sent.GetMessageId(),
		Payload: &memqlv1.MemqlServerMessage_QueryResult{
			QueryResult: &memqlv1.QueryResultChunk{
				Result: &memqlv1.Result{
					Bundle: &memqlv1.GraphBundle{
						Nodes: []*memqlv1.MemoryNode{{
							Id:      "v1:cognition:utterance:u-a",
							Concept: "v1:cognition:utterance",
							Payload: payload,
						}},
					},
					Meta: &memqlv1.ResultMeta{Cursor: "cursor-to-page-2", HasMore: true},
				},
			},
		},
	}

	select {
	case got := <-done:
		if got.err != nil {
			t.Fatalf("BrowseConceptPage: %v", got.err)
		}
		if got.page.NextCursor != "cursor-to-page-2" {
			t.Errorf("NextCursor: want cursor-to-page-2, got %q", got.page.NextCursor)
		}
		if !got.page.HasMore {
			t.Errorf("HasMore: want true")
		}
		if len(got.page.Rows) != 1 {
			t.Fatalf("rows: want 1, got %d", len(got.page.Rows))
		}
		if RowString(got.page.Rows[0], "id") != "v1:cognition:utterance:u-a" {
			t.Errorf("row id intrinsic not carried: %+v", got.page.Rows[0])
		}
		if p, ok := got.page.Rows[0]["payload"].(map[string]any); !ok || p["text"] != "hello" {
			t.Errorf("nested payload dropped (cockpit inspector needs it): %+v", got.page.Rows[0])
		}
	case <-time.After(2 * time.Second):
		t.Fatal("timed out waiting for BrowseConceptPage to return")
	}
}

// TestBrowseConceptPage_DefaultPageSize proves a non-positive pageSize falls
// back to DefaultConceptBrowsePageSize in the emitted paginate() literal.
func TestBrowseConceptPage_DefaultPageSize(t *testing.T) {
	stream := newMockStream()
	d := NewDispatcher(stream, nil)
	go d.Run()
	defer d.Stop()
	qc := NewQueryClient(d)

	go func() { _, _ = qc.BrowseConceptPage(context.Background(), "v1:cluster:node", "", 0) }()

	sent := <-stream.sendCh
	q := sent.GetExecuteQuery()
	wantQuery := `sort(paginate(concept==v1:cluster:node, 200), "createdAt", "asc")`
	if q.GetQuery() != wantQuery {
		t.Errorf("query: want %q, got %q (pageSize<=0 must use DefaultConceptBrowsePageSize=%d)", wantQuery, q.GetQuery(), DefaultConceptBrowsePageSize)
	}
	if q.GetCursor() != "" {
		t.Errorf("cursor: want empty on first page, got %q", q.GetCursor())
	}
	// Drain so the dispatcher waiter completes cleanly.
	stream.recvCh <- &memqlv1.MemqlServerMessage{
		CorrelateTo: sent.GetMessageId(),
		Payload: &memqlv1.MemqlServerMessage_QueryResult{
			QueryResult: &memqlv1.QueryResultChunk{Result: &memqlv1.Result{}},
		},
	}
}

// TestBrowseConcept_RequiresConceptId guards the empty-id error on both forms.
func TestBrowseConcept_RequiresConceptId(t *testing.T) {
	qc := NewQueryClient(nil)
	if _, err := qc.BrowseConcept(context.Background(), ""); err == nil {
		t.Error("BrowseConcept(\"\"): want error")
	}
	if _, err := qc.BrowseConceptPage(context.Background(), "", "", 0); err == nil {
		t.Error("BrowseConceptPage(\"\"): want error")
	}
}

// TestGetRowByConceptAndId_UsesCanonicalAndConnective pins the filter this
// SDK emits for a single-row lookup.
//
// The two clauses used to be joined with the legacy `;`-AND separator, which
// memql#971 retired in favour of the one Go boolean grammar -- and which the
// TS SDK's getRowByConceptAndId had already moved to. Nothing downstream
// noticed, because both spellings still parse; that is exactly why it needs a
// test rather than a comment. Two SDKs spelling the same two-clause filter
// differently is a difference every reader has to stop and confirm is not
// meaningful.
func TestGetRowByConceptAndId_UsesCanonicalAndConnective(t *testing.T) {
	stream := newMockStream()
	d := NewDispatcher(stream, nil)
	go d.Run()
	defer d.Stop()
	qc := NewQueryClient(d)

	go func() { _, _ = qc.GetRowByConceptAndId(context.Background(), "v1:agents:agent", "a1") }()

	sent := <-stream.sendCh
	q := sent.GetExecuteQuery()
	if q == nil {
		t.Fatalf("expected ExecuteQueryMsg payload, got %+v", sent.GetPayload())
	}
	wantQuery := "concept==v1:agents:agent && id==a1"
	if q.GetQuery() != wantQuery {
		t.Errorf("query: want %q, got %q (the retired `;`-AND form must not come back)", wantQuery, q.GetQuery())
	}

	// Drain so the dispatcher waiter completes cleanly.
	stream.recvCh <- &memqlv1.MemqlServerMessage{
		CorrelateTo: sent.GetMessageId(),
		Payload: &memqlv1.MemqlServerMessage_QueryResult{
			QueryResult: &memqlv1.QueryResultChunk{Result: &memqlv1.Result{}},
		},
	}
}

// TestGetRowByConceptAndId_RequiresBothIds guards the empty-argument errors.
func TestGetRowByConceptAndId_RequiresBothIds(t *testing.T) {
	qc := NewQueryClient(nil)
	if _, err := qc.GetRowByConceptAndId(context.Background(), "", "a1"); err == nil {
		t.Error("GetRowByConceptAndId(\"\", \"a1\"): want error")
	}
	if _, err := qc.GetRowByConceptAndId(context.Background(), "v1:agents:agent", ""); err == nil {
		t.Error("GetRowByConceptAndId(concept, \"\"): want error")
	}
}
